src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,433 @@
+        "com.puppycrawl.tools.checkstyle.checks.naming.ParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.PatternVariableNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.RecordComponentNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.RecordTypeParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.StaticVariableNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.TypeNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.whitespace.TypecastParenPadCheck",
+        // Failing 'testClassesWithoutImmutableFieldsAreStateful'
+        "com.puppycrawl.tools.checkstyle.checks.whitespace.ParenPadCheck"
+    );
+
+    /**
+     * ArchCondition for immutable fields.
+     */
+    private static final ArchCondition<JavaField> BE_IMMUTABLE = new ImmutableFieldArchCondition();
+
+    /**
+     * Map fo module full name to module details.
+     */
+    private static final Map<String, ModuleDetails> MODULE_DETAILS_MAP =
+        XmlMetaReader.readAllModulesIncludingThirdPartyIfAny().stream()
+            .collect(Collectors.toMap(ModuleDetails::getFullQualifiedName,
+                                      moduleDetails -> moduleDetails));
+
+    /**
+     * Test to ensure that fields in util classes are immutable.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testUtilClassesImmutability() {
+        final JavaClasses utilClasses = new ClassFileImporter()
+            .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
+            .importPackages("com.puppycrawl.tools.checkstyle.utils",
+                            "com.puppycrawl.tools.checkstyle.checks.javadoc.utils");
+
+        final DescribedPredicate<JavaField> zeroSizeArrays =
+            new DescribedPredicate<>("zero size arrays") {
+                @Override
+                public boolean apply(JavaField input) {
+                    return ZERO_SIZE_ARRAY_FIELDS.contains(input.getFullName());
+                }
+            };
+
+        final ArchCondition<JavaField> beSuppressed = new ArchCondition<>("be suppressed") {
+            @Override
+            public void check(JavaField item, ConditionEvents events) {
+                if (!SUPPRESSED_FIELDS_IN_UTIL_CLASSES.contains(item.getFullName())) {
+                    final String message = String
+                        .format(Locale.ROOT, "should %s or resolved.", getDescription());
+                    events.add(SimpleConditionEvent.violated(item, message));
+                }
+            }
+        };
+
+        final ArchRule fieldsShouldBeImmutable = fields()
+            .that(are(not(zeroSizeArrays)))
+            .and()
+            .areDeclaredInClassesThat()
+            .haveSimpleNameEndingWith("Util")
```

Comment:
Is it possible to get all classes by package, as for api/util package rule?

Reply:
We are getting the required classes only, the ones present in `utils` packages, but there are some classes we need to exclude, like `com.puppycrawl.tools.checkstyle.checks.javadoc.utils.TagInfo` that's why this condition is there.